### ISSUE TITLE
Add SUM() function to pg_vector

### ISSUE THREAD

**Issue by ThierDev** [pgvector/pgvector#160]
The SUM() function would be a valuable addition to the pg_vector extension. We are utilizing the lightfm library for recommendations. This models estimate latent vectors not only for users and items but also for their respective metadata. Quoting lightfm "The latent representation of user/item u is given by the sum of its features’ latent vectors/embedding" https://arxiv.org/pdf/1507.08439.pdf

In that context the SUM() function would be extremely useful. For instance, consider the representation of an item called "Dior Evening Dress." This representation would involve the element-wise sum of the embeddings for the words "Dior," "Evening," and "Dress." and not the average actually. 

Currently, the extension offers the AVG() function, we did it on our fork, it took us around 5 min to copy/paste AVG and remove the /n , thus giving us the SUM(). We could push that as pull request, let us know. 

We need this as part of an official releases because we use the extensions on AMAZON RDS for POSTGRES. 

We kindly request you to consider adding this function to the next release of pg_vector.

**Comment by ankane** [comment]
Hi @ThierDev, thanks for sharing! Can you share the specific SQL queries you're using for it?

Also, fwiw, I think it might be difficult to do efficient search on the model even if the user and item representations are materialized since the prediction involves biases (so won't fully be able to be indexed).

**Comment by m-gris** [comment]
Hi @ankane  

After giving it a second thought, @ThierDev  and I realized that given our use-case, we could indeed simply work with AVG instead of SUM, since their respective outputs would be colinear... 

Thanks anyhow.  
Regards 
M.

**Comment by jkatz** [comment]
@ThierDev @m-gris If you're still looking for a way to do `SUM` on RDS, you can use PL/Rust to implement this. Here are [some instructions](https://aws.amazon.com/blogs/database/build-high-performance-functions-in-rust-on-amazon-rds-for-postgresql/) for enabling the appropriate parameter on your instance, and an example function:

```sql
-- function used to generate test data
CREATE OR REPLACE FUNCTION generate_random_vector(dim int)
RETURNS float4[]
LANGUAGE plpgsql
STRICT
AS $$
    DECLARE
        i int := 1;
        v float4[];
    BEGIN
        FOR i IN 1..dim LOOP
            v[i] := random();
        END LOOP;

        RETURN v;
    END;
$$;

-- create the extensions
CREATE EXTENSION IF NOT EXISTS plrust;
CREATE EXTENSION IF NOT EXISTS vector;

-- generate test data
CREATE TABLE vectors (
    id int PRIMARY KEY,
    embedding vector(768)
);

INSERT INTO vectors
SELECT n, generate_random_vector(768)
FROM generate_series (1,1000000) n;

-- prior to PL/Rust 1.2.0, you can use this function to create the accumulator method:
CREATE OR REPLACE FUNCTION public.vector_add(x float4[], y float4[])
RETURNS float4[]
LANGUAGE plrust
IMMUTABLE PARALLEL SAFE STRICT
AS $$
    match x.len() {
        0 => {
            Ok(Some(y.into_iter().collect()))
        }
        _ => {
            let mut iter = y.into_iter().zip(x.into_iter());

            let z: Vec<Option<f32>> = iter.map(|(xi, yi)| Some(xi.unwrap() + yi.unwrap()) ).collect();

            Ok(Some(z))
        }
    }
$$;

-- With PL/Rust 1.2.0 and higher, you can use the following function, which can be ~25% more performant:
CREATE OR REPLACE FUNCTION public.vector_add(x float4[], y float4[])
    RETURNS float4[]
    LANGUAGE plrust
    IMMUTABLE PARALLEL SAFE STRICT
AS $$
    match x.len() {
        0 => {
            Ok(Some(y.into_iter().collect()))
        }
        _ => {
            let z = y.as_slice().into_iter()
                .zip(
                    x.as_slice().into_iter()
                )
                .map(|(xi, yi)| Some(xi + yi))
                .collect();

            Ok(Some(z))
        }
    }
$$;

-- this creates the custom aggreagtor:
CREATE OR REPLACE AGGREGATE vector_agg(float4[])
(
    sfunc = vector_add,
    stype = float4[],
    initcond = '{}'
);

-- test!
EXPLAIN ANALYZE SELECT vector_agg(embedding)
FROM vectors;
```

The two key pieces to create the custom aggregate are the following:

```sql
CREATE OR REPLACE FUNCTION public.vector_add(x float4[], y float4[])
RETURNS float4[]
LANGUAGE plrust
IMMUTABLE PARALLEL SAFE STRICT
AS $$
    match x.len() {
        0 => {
            Ok(Some(y.into_iter().collect()))
        }
        _ => {
            let mut iter = y.into_iter().zip(x.into_iter());

            let z: Vec<Option<f32>> = iter.map(|(xi, yi)| Some(xi.unwrap() + yi.unwrap()) ).collect();

            Ok(Some(z))
        }
    }
$$;

CREATE OR REPLACE AGGREGATE vector_agg(float4[])
(
    sfunc = vector_add,
    stype = float4[],
    initcond = '{}'
);
```

To compare, I took the example from https://github.com/pgvector/pgvector/issues/141#issuecomment-1594334001 and rewrote it to be more optimal (shown below). In my tests, I found the PL/Rust equivalent was 2-3x more performant.

```sql
CREATE OR REPLACE FUNCTION vector_add_plpgsql(_a float4[], _b float4[])
RETURNS float4[]
AS
$$
BEGIN
    IF array_length(_a, 1) = 0 THEN
        RETURN _b;
    END IF;

    RETURN ARRAY(
    SELECT coalesce(a, 0) + coalesce(b, 0)
    FROM unnest(_a, _b) WITH ORDINALITY AS x(a, b, n)
    ORDER BY n
);
END
$$ LANGUAGE plpgsql;

CREATE OR REPLACE AGGREGATE vector_agg_plpgsql(float4[])
(
    sfunc = vector_add_plpgsql,
    stype = float4[],
    initcond = '{}'
);
```

Hope this helps!

**Comment by m-gris** [comment]
@jkatz  Thx a lot for sharing this 🙏 😊

**Comment by e253** [comment]
Simplest solution, pulled from [#13](https://github.com/pgvector/pgvector/pull/13). The vector_add function is already defined in vector.c https://github.com/pgvector/pgvector/blob/master/src/vector.c#L685
```
CREATE AGGREGATE sum(vector)
(
    sfunc=vector_add,
    stype = vector,
    combinefunc = vector_add,
    parallel = safe
);
```

**Comment by ankane** [comment]
Just fyi, this was added in ed1bc3e856d1726426217c2e71e8681a47856f55 and will be included in 0.5.0 (if it's still needed for anything). Thanks again for sharing the paper and use case.
